moonraker/components/wled.py

```diff
@@ -0,0 +1,445 @@
+
+            # Register two remote methods for GCODE
+            self.server.register_remote_method(
+                "set_wled_state", self.set_wled_state)
+            self.server.register_remote_method(
+                "set_wled", self.set_wled)
+
+            # As moonraker is about making things a web api, let's try it
+            # Yes, this is largely a cut-n-paste from power.py
+            self.server.register_endpoint(
+                "/machine/wled/strips", ["GET"],
+                self._handle_list_strips)
+            self.server.register_endpoint(
+                "/machine/wled/status", ["GET"],
+                self._handle_batch_wled_request)
+            self.server.register_endpoint(
+                "/machine/wled/on", ["POST"],
+                self._handle_batch_wled_request)
+            self.server.register_endpoint(
+                "/machine/wled/off", ["POST"],
+                self._handle_batch_wled_request)
+            self.server.register_endpoint(
+                "/machine/wled/strip", ["GET", "POST"],
+                self._handle_single_wled_request)
+
+        except Exception as e:
+            logging.exception(e)
+
+    async def _initalize_strips(self: WLED,
+                                initial_strips: List[Strip]
+                                ) -> None:
+        try:
+            logging.debug("Initializing wled")
+            event_loop = self.server.get_event_loop()
+            cur_time = event_loop.get_loop_time()
+            endtime = cur_time + 120.
+            query_strips = initial_strips
+            failed_strips: List[Strip] = []
+            while cur_time < endtime:
+                for strip in query_strips:
+                    ret = strip.initialize()
+                    if ret is not None:
+                        await ret
+                    if strip.error_state is not None:
+                        failed_strips.append(strip)
+                if not failed_strips:
+                    logging.debug("All wled strips initialized")
+                    return
+                query_strips = failed_strips
+                failed_strips = []
+                await asyncio.sleep(2.)
+                cur_time = event_loop.get_loop_time()
+            if failed_strips:
+                failed_names = [s.name for s in failed_strips]
+                self.server.add_warning(
+                    "The following wled strips failed init:"
+                    f" {failed_names}")
+        except Exception as e:
+            logging.exception(e)
+
+    async def wled_on(self: WLED, strip: str, preset: int) -> None:
+        if strip not in self.strips:
+            logging.info(f"Unknown WLED strip: {strip}")
+            return
+        await self.strips[strip].wled_on(preset)
+
+    # Full control of wled, True, False, "on", "off", preset 42 as int or
+    # preset "42" as string
+    async def set_wled_state(self: WLED, strip: str, state: str) -> None:
```

Comment:
I think it would make sense to have an additional "preset" argument here that defaults to -1 rather than attempt to use the state for both, ie:

```python 
async def set_wled_state(self: WLED, strip: str, state: str, preset: int = -1) -> None:
```

Reply:
Misunderstood your previous suggestion, fixed up